core/unix/os.c

```diff
@@ -9420,11 +9434,16 @@ mutex_wait_contended_lock(mutex_t *lock)
 #ifdef CLIENT_INTERFACE
             if (set_client_safe_for_synch)
                 dcontext->client_data->client_thread_safe_for_synch = true;
+            if (mc != NULL)
+                set_synch_state(dcontext, THREAD_SYNCH_VALID_MCONTEXT);
 #endif
+
             os_thread_yield();
 #ifdef CLIENT_INTERFACE
             if (set_client_safe_for_synch)
                 dcontext->client_data->client_thread_safe_for_synch = false;
+            if (mc != NULL)
```

Comment:
Seems safest to never support both of these if()'s being true: relocating the client thread case is not going to work as nobody will reset client_thread_safe_for_synch.

Reply:
Adding an assert at the top of the function (both here in win32/os.c)